### PR TITLE
Add stubs

### PR DESCRIPTION
**Pull request by flying-sheep**
TODO:

- [x] figure out how to automatically generate the stubs when building

Eventually we should also get https://github.com/Jij-Inc/pyo3-stub-gen/pull/102 merged and released, but that’s not important, as we will publish to PyPI and not to crates.io

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -5,8 +5,8 @@ edition = "2021"
 
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 [lib]
-name = "zarrs"
```
Comment (flying-sheep) [inline Cargo.toml:8]: I wonder why that worked. In #37, you made it so there are two Rust libraries called `zarrs`, which should be impossible.

**Comment by flying-sheep** [comment]
The generated file looks like this:

```py
# This file is automatically generated by pyo3_stub_gen
# ruff: noqa: E501, F401

import numpy
import numpy.typing
import typing

class CodecPipelineImpl:
    def __new__(cls,metadata,validate_checksums = ...,store_empty_chunks = ...,concurrent_target = ...): ...
    def retrieve_chunks_and_apply_index(self, chunk_descriptions:typing.Sequence[tuple[tuple[str, typing.Sequence[int], str, typing.Sequence[int]], typing.Sequence[slice], typing.Sequence[slice]]], value:numpy.NDArray[typing.Any], chunk_concurrent_limit:int) -> None:
        ...

    def retrieve_chunks(self, chunk_descriptions:typing.Sequence[tuple[str, typing.Sequence[int], str, typing.Sequence[int]]], chunk_concurrent_limit:int) -> list[numpy.typing.NDArray[numpy.uint8]]:
        ...

    def store_chunks_with_indices(self, chunk_descriptions:typing.Sequence[tuple[tuple[str, typing.Sequence[int], str, typing.Sequence[int]], typing.Sequence[slice], typing.Sequence[slice]]], value:numpy.NDArray[typing.Any], chunk_concurrent_limit:int) -> None:
        ...
```

I assume that the type inference for `__new__` doesn‘t work because we use `#[pyo3(signature=...)]`.

File: Cargo.toml (hunk 1 above)
Comment (ilan-gold) [reply Cargo.toml:8]: That's a good point.

**Review by ilan-gold — COMMENTED** [review]
Could you add to docs how to use this? In `contributing.md`?

**Review by LDeakin — APPROVED** [review]

**Pull request merged**